File: webui-src/src/api/crawler.ts
// 爬虫 API 接口
import { client } from './client'

// 类型定义（对应后端的Schema）
export type Platform = 'xhs' | 'dy' | 'ks' | 'bili' | 'wb' | 'wechat' | 'tieba' | 'zhihu'
export type LoginType = 'qrcode' | 'phone' | 'cookie' | 'mp_qrcode'
export type CrawlerType = 'search' | 'detail' | 'creator' | 'creator_vip' | 'album'
export type SaveOption = 'json' | 'csv' | 'db' | 'sqlite' | 'mongodb' | 'excel'

export interface CrawlerStartRequest {
  platform: Platform
  login_type: LoginType
  crawler_type: CrawlerType
  keywords?: string
  specified_ids?: string
  creator_ids?: string
  vip_creator_ids?: string
  start_page?: number
  enable_comments?: boolean
  enable_sub_comments?: boolean
  save_option?: SaveOption
  cookies?: string
  headless?: boolean
  // 增量爬取配置
  enable_incremental?: boolean  // 是否启用增量爬取
  incremental_early_stop?: number  // 早停阈值
  // 微信专用配置
  wechat_enable_content?: boolean  // 是否下载文章HTML内容
  wechat_enable_reading_stats?: boolean  // 是否获取阅读量
  wechat_enable_export?: boolean  // 是否启用导出
  wechat_export_format?: 'html' | 'markdown' | 'txt' | 'docx' | 'json' | 'excel'  // 导出格式
  wechat_album_ids?: string  // 合集ID列表
  wechat_credentials_uin?: string  // 微信凭证 uin
  wechat_credentials_key?: string  // 微信凭证 key
  wechat_credentials_pass_ticket?: string  // 微信凭证 pass_ticket
  wechat_token?: string // 微信后台 token
  // 仅登录模式
  login_only?: boolean  // 是否只登录获取Cookie/Token，不进行数据爬取
}

export interface CrawlerStatus {
  status: 'idle' | 'running' | 'stopping' | 'error'
  platform?: string
  crawler_type?: string
  started_at?: string
  error_message?: string
  // 进度信息
  progress?: CrawlerProgress
  new_cookies?: string // 登录成功后获取的新Cookie
  new_token?: string // 登录成功后获取的新Token
  qrcode_img?: string // 扫码登录的二维码 (base64)
}

export interface CrawlerProgress {
  articles_crawled?: number
  articles_total?: number
  comments_crawled?: number
  resources_downloaded?: number
  exports_completed?: number
  current_account?: string
  current_album?: string
  percentage?: number
}

export interface LogEntry {
  id: number
  timestamp: string
  level: 'info' | 'warning' | 'error' | 'success' | 'debug'
  message: string
}

// 配置选项类型
export interface PlatformOption {
  value: string
  label: string
  icon: string
}

export interface ConfigOption {
  value: string
  label: string
}

export interface IncrementalConfig {
  description: string
  supports_platforms: Platform[]
  supports_crawler_types: CrawlerType[]
  default_enabled: boolean
  default_threshold: number
  threshold_range: {
    min: number
    max: number
  }
}

export interface ConfigOptions {
  login_types: ConfigOption[]
  crawler_types: ConfigOption[]
  save_options: ConfigOption[]
  incremental_config?: IncrementalConfig
}

// 数据文件类型
export interface DataFile {
  name: string
  path: string
  size: number
  modified_at: number
  record_count: number | null
  type: string
}

// 微信文章类型
export interface WeChatArticleItem {
  id: number
  article_id: string
  title: string
  account_name: string
  read_num: number
  like_num: number
  comment_count: number
  create_time: number
  link: string
  cover?: string
}

// 微信文章列表响应
export interface WeChatArticleListResponse {
  articles: WeChatArticleItem[]
  total: number
  page: number
  page_size: number
}

// 微信统计数据
export interface WeChatStats {
  total_articles: number
  total_reads: number
  total_likes: number
  total_accounts: number
  today_articles: number
  today_reads: number
}

// 热门文章
export interface WeChatTopArticle {
  id: number
  title: string
  account_name: string
  read_num: number
  create_time: number
}

// API 方法
export const crawlerApi = {
  // 启动爬虫
  start: (data: CrawlerStartRequest) => 
    client.post<void>('/api/crawler/start', data),
  
  // 停止爬虫
  stop: () => 
    client.post<void>('/api/crawler/stop'),
  
  // 获取状态
  getStatus: () => 
    client.get<CrawlerStatus>('/api/crawler/status'),
  
  // 获取日志
  getLogs: (limit = 100) => 
    client.get<{ logs: LogEntry[] }>(`/api/crawler/logs?limit=${limit}`),
  
  // 获取平台列表
  getPlatforms: () => 
    client.get<{ platforms: PlatformOption[] }>('/api/config/platforms'),
  
  // 获取配置选项
  getConfigOptions: () => 
    client.get<ConfigOptions>('/api/config/options'),
  
  // 获取数据文件列表
  getDataFiles: (platform?: string, fileType?: string) => {
    const params = new URLSearchParams()
    if (platform) params.append('platform', platform)
    if (fileType) params.append('file_type', fileType)
    return client.get<{ files: DataFile[] }>(`/api/data/files?${params.toString()}`)
  },
  
  // 下载数据文件
  downloadFile: (filePath: string) => {
    return `/api/data/download/${filePath}`
  },
  
  // 微信公众号搜索 (需登录)
  searchWeChatAccount: (keyword: string, cookies: string, token: string) => {
    return client.post<{ list: any[] }>('/api/wechat/search_account', { 
      keyword, 
      cookies, 
      token 
    })
  },
  
  // 获取微信文章列表
  getWeChatArticles: (params: {
    page?: number
    page_size?: number
    search?: string
    time_range?: 'today' | 'week' | 'month' | 'all'
    order_by?: 'create_time' | 'read_num' | 'like_num'
    order_dir?: 'asc' | 'desc'
  } = {}) => {
    const queryParams = new URLSearchParams()
    if (params.page) queryParams.append('page', params.page.toString())
    if (params.page_size) queryParams.append('page_size', params.page_size.toString())
    if (params.search) queryParams.append('search', params.search)
    if (params.time_range) queryParams.append('time_range', params.time_range)
    if (params.order_by) queryParams.append('order_by', params.order_by)
    if (params.order_dir) queryParams.append('order_dir', params.order_dir)
    return client.get<WeChatArticleListResponse>(`/api/wechat/articles?${queryParams.toString()}`)
  },
  
  // 获取微信统计数据
  getWeChatStats: () => {
    return client.get<WeChatStats>('/api/wechat/stats')
  },
  
  // 获取热门文章
  getWeChatTopArticles: (limit = 5) => {
    return client.get<WeChatTopArticle[]>(`/api/wechat/top_articles?limit=${limit}`)
  },
}

File: api/schemas/crawler.py
# -*- coding: utf-8 -*-
# Copyright (c) 2025 relakkes@gmail.com
#
# This file is part of MediaCrawler project.
# Repository: https://github.com/NanmiCoder/MediaCrawler/blob/main/api/schemas/crawler.py
# GitHub: https://github.com/NanmiCoder
# Licensed under NON-COMMERCIAL LEARNING LICENSE 1.1
#
# 声明：本代码仅供学习和研究目的使用。使用者应遵守以下原则：
# 1. 不得用于任何商业用途。
# 2. 使用时应遵守目标平台的使用条款和robots.txt规则。
# 3. 不得进行大规模爬取或对平台造成运营干扰。
# 4. 应合理控制请求频率，避免给目标平台带来不必要的负担。
# 5. 不得用于任何非法或不当的用途。
#
# 详细许可条款请参阅项目根目录下的LICENSE文件。
# 使用本代码即表示您同意遵守上述原则和LICENSE中的所有条款。

from enum import Enum
from typing import Optional, Literal
from pydantic import BaseModel


class PlatformEnum(str, Enum):
    """Supported media platforms"""
    XHS = "xhs"
    DOUYIN = "dy"
    KUAISHOU = "ks"
    BILIBILI = "bili"
    WEIBO = "wb"
    WECHAT = "wechat"
    TIEBA = "tieba"
    ZHIHU = "zhihu"


class LoginTypeEnum(str, Enum):
    """Login method"""
    QRCODE = "qrcode"
    PHONE = "phone"
    COOKIE = "cookie"
    MP_QRCODE = "mp_qrcode"  # WeChat MP backend login


class CrawlerTypeEnum(str, Enum):
    """Crawler type"""
    SEARCH = "search"
    DETAIL = "detail"
    CREATOR = "creator"
    CREATOR_VIP = "creator_vip"  # VIP exclusive content from creators (Weibo only)
    ALBUM = "album"  # WeChat album mode


class SaveDataOptionEnum(str, Enum):
    """Data save option"""
    CSV = "csv"
    DB = "db"
    JSON = "json"
    SQLITE = "sqlite"
    MONGODB = "mongodb"
    EXCEL = "excel"


class ExportFormatEnum(str, Enum):
    """Export format options"""
    HTML = "html"
    MARKDOWN = "markdown"
    TXT = "txt"
    DOCX = "docx"
    JSON = "json"
    EXCEL = "excel"


class CrawlerStartRequest(BaseModel):
    """Crawler start request"""
    platform: PlatformEnum
    login_type: LoginTypeEnum = LoginTypeEnum.QRCODE
    crawler_type: CrawlerTypeEnum = CrawlerTypeEnum.SEARCH
    keywords: str = ""  # Keywords for search mode
    specified_ids: str = ""  # Post/video ID list for detail mode, comma-separated
    creator_ids: str = ""  # Creator ID list for creator mode, comma-separated
    vip_creator_ids: str = ""  # VIP creator ID list for creator_vip mode (Weibo only), comma-separated
    start_page: int = 1
    enable_comments: bool = True
    enable_sub_comments: bool = False
    save_option: SaveDataOptionEnum = SaveDataOptionEnum.JSON
    cookies: str = ""
    headless: bool = False
    # 增量爬取配置
    enable_incremental: bool = False  # 是否启用增量爬取（只爬取新内容）
    incremental_early_stop: int = 3  # 早停阈值（连续N条已存在内容就停止）
    
    # 微信专用配置
    wechat_enable_content: bool = False  # 是否下载文章HTML内容
    wechat_enable_reading_stats: bool = False  # 是否获取阅读量
    wechat_enable_export: bool = False  # 是否启用导出
    wechat_export_format: Optional[str] = "html"  # 导出格式
    wechat_album_ids: str = ""  # 合集ID列表，用于album模式
    wechat_credentials_uin: str = ""  # 微信凭证 uin
    wechat_credentials_key: str = ""  # 微信凭证 key
    wechat_credentials_pass_ticket: str = ""  # 微信凭证 pass_ticket
    wechat_token: str = ""  # 微信后台 token
    
    # 仅登录模式
    login_only: bool = False  # 是否只登录获取Cookie/Token，不进行数据爬取


class CrawlerStatusResponse(BaseModel):
    """Crawler status response"""
    status: Literal["idle", "running", "stopping", "error"]
    platform: Optional[str] = None
    crawler_type: Optional[str] = None
    started_at: Optional[str] = None
    error_message: Optional[str] = None
    # 进度信息
    progress: Optional[dict] = None  # 包含 current, total, percentage 等
    new_cookies: Optional[str] = None  # 登录成功后获取的新Cookie
    new_token: Optional[str] = None  # 登录成功后获取的新Token
    qrcode_img: Optional[str] = None  # 扫码登录的二维码 (base64)


class WeChatProgressInfo(BaseModel):
    """WeChat crawler progress information"""
    articles_crawled: int = 0
    articles_total: int = 0
    comments_crawled: int = 0
    resources_downloaded: int = 0
    exports_completed: int = 0
    current_account: Optional[str] = None
    current_album: Optional[str] = None


class LogEntry(BaseModel):
    """Log entry"""
    id: int
    timestamp: str
    level: Literal["info", "warning", "error", "success", "debug"]
    message: str


class DataFileInfo(BaseModel):
    """Data file information"""
    name: str
    path: str
    size: int
    modified_at: str
    record_count: Optional[int] = None
